internal/ark/api/backups.go

```diff
@@ -0,0 +1,170 @@
+// Copyright © 2018 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	arkAPI "github.com/heptio/ark/pkg/apis/ark/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	// LabelKeyCloud label key used for cloud information
+	LabelKeyCloud = "pipeline-cloud"
+	// LabelKeyDistribution label key used for distribution information
+	LabelKeyDistribution = "pipeline-distribution"
+	// LabelKeyNodeCount label key is used for node count
+	LabelKeyNodeCount = "pipeline-nodecount"
+)
+
+// PersistBackupRequest describes a backup persisting request
+type PersistBackupRequest struct {
+	BucketID uint
+
+	Cloud          string
+	Distribution   string
+	NodeCount      uint
+	ContentChecked bool
+
+	ClusterID    uint
+	DeploymentID uint
+
+	Nodes  *core.NodeList
+	Backup *arkAPI.Backup
+}
+
+// Backup describes an ARK backup
+type Backup struct {
+	UID              string                              `json:"uid"`
+	Name             string                              `json:"name"`
+	TTL              metav1.Duration                     `json:"ttl"`
+	Labels           labels.Set                          `json:"labels"`
+	Cloud            string                              `json:"cloud"`
+	Distribution     string                              `json:"distribution"`
+	Options          BackupOptions                       `json:"options,omitempty"`
+	Status           string                              `json:"status"`
+	StartAt          time.Time                           `json:"startAt"`
+	ExpireAt         time.Time                           `json:"expireAt"`
+	VolumeBackups    map[string]*arkAPI.VolumeBackupInfo `json:"volumeBackups,omitempty"`
+	ValidationErrors []string                            `json:"validationErrors,omitempty"`
+
+	ClusterID uint    `json:"clusterId,omitempty"`
+	Bucket    *Bucket `json:"-"`
+}
+
+// DeleteBackupResponse describes a delete backup response
+type DeleteBackupResponse struct {
+	Name   string `json:"name"`
+	Status int    `json:"status"`
+}
+
+// CreateBackupRequest descibes a create backup request
+type CreateBackupRequest struct {
+	Name    string          `json:"name" binding:"required"`
+	TTL     metav1.Duration `json:"ttl" binding:"required"`
```

Comment:
Are you sure this is required? I think this should be optional

Reply:
I think it is an important field and although there is a default for it, it maybe better if the caller must set it explicitly.